contracts/src/L1/L1BridgeController.sol

```
@@ -0,0 +1,207 @@
+            _assertTokenIdMatchesLabel(tokenId, transferData.label);
+            
+            // burn the token
+            registry.burn(tokenId);
+
+            // send the message to the bridge
+            bytes memory dnsEncodedName = NameUtils.dnsEncodeEthLabel(transferData.label);
+            bridge.sendMessage(BridgeEncoder.encodeEjection(dnsEncodedName, transferData));
+            emit NameEjectedToL2(dnsEncodedName, tokenId);
+        }
+    }
+
+    /**
+     * @dev Registers a name in the specified registry.
+     *
+     * @param transferData The transfer data for the name being registered
+     * @param targetRegistry The registry to register the name in
+     */
+    function _registerName(
+        TransferData memory transferData,
+        IPermissionedRegistry targetRegistry
+    ) private returns (uint256 tokenId) {
+        tokenId = targetRegistry.register(
+            transferData.label,
+            transferData.owner,
+            IRegistry(transferData.subregistry),
+            transferData.resolver,
+            transferData.roleBitmap,
+            transferData.expires
+        );
+    }
+
+    /**
+     * @dev Finds the parent registry for a DNS-encoded name by traversing the hierarchy.
+     * For a name like "sub.parent", this will:
+     * 1. Find "parent" in root registry
+     * 2. Return parent's registry so "sub" can be registered there
+     * Reverts if any parent component is not found in the registry.
+     *
+     * @param dnsEncodedName The DNS-encoded name
+     * @param offset The current offset in the name
+     */
+    function _findParentRegistryForMigration(
+        bytes memory dnsEncodedName,
+        uint256 offset
+    ) private view returns (IRegistry) {
+        // Read the current label (e.g., "sub" in "sub.parent")
+        (bytes32 labelHash, uint256 nextOffset) = NameCoder.readLabel(dnsEncodedName, offset);
+        
+        // If we're at the end (null terminator), something is wrong
```

Surely we shouldn't silently return the default value, then?